### ISSUE TITLE
[BUG] Object detection class head is not re-initialized when the number of classes changes

### ISSUE THREAD

**Issue by liopeer** [lightly-ai/lightly-train#982]
### 🧠 Describe the Bug

Fine-tuning object detection from a COCO-pretrained checkpoint onto a different number of classes does **not** re-initialize the classification head. It silently keeps the checkpoint's leading rows.

`_reuse_or_reinit` in `src/lightly_train/_task_models/object_detection_components/hooks.py:168` truncates when the checkpoint has more classes:

```python
if num_classes_checkpoint > num_classes_module:
    # Checkpoint has more classes: truncate to module's expected size.
    adjusted_weights = checkpoint_weight[:num_classes_module, :].to(device)
```

So fine-tuning COCO (80 classes) onto 3 classes starts from the COCO rows for `person`, `bicycle` and `car`. Training still converges, but from an arbitrary bias that depends on COCO's class ordering.

This contradicts:

- The `train_object_detection` docstring: *"Otherwise, the classification head will be re-initialized."*
- The module docstring, which claims the hooks work by "truncating or zero-/re-initializing".
- `src/lightly_train/_task_models/eomt/hooks.py:85`, where the segmentation tasks already re-initialize correctly.

`denoising_class_embed_reuse_or_reinit_hook` (same file, line 31) has the same problem.

### 🔁 Steps to Reproduce

```python
import torch

from lightly_train._task_models import task_model_helpers
from lightly_train._task_models.ltdetr_object_detection.task_model import (
    LTDETRObjectDetection,
)

ckpt = torch.load(
    task_model_helpers.download_checkpoint(checkpoint="ltdetrv2-s-coco"),
    weights_only=False,
    map_location="cpu",
)
model = LTDETRObjectDetection(
    **{**ckpt["model_init_args"], "classes": {0: "cat"}, "load_weights": False}
)

fresh = model.decoder.dec_score_head[0].weight.detach().clone()
model.load_train_state_dict(state_dict=ckpt["train_model"])
after = model.decoder.dec_score_head[0].weight
coco = ckpt["train_model"]["ema_model.model.decoder.dec_score_head.0.weight"]

print("head was re-initialized:", torch.equal(fresh, after))
print("head equals COCO row 0 ('person'):", torch.equal(after[0], coco[0]))
```

Output:

```
head was re-initialized: False
head equals COCO row 0 ('person'): True
```

### 📌 Expected Behavior

When the class counts differ, keep the module's own initialization rather than reusing checkpoint rows, the way `eomt/hooks.py` already does:

```python
state_dict[weight_key] = head_module.weight.detach().clone()
state_dict[bias_key] = head_module.bias.detach().clone()
```

When the counts match, reuse the checkpoint as before.

### 🛠️ Scope

Small and self-contained:

- Fix `_reuse_or_reinit` and `denoising_class_embed_reuse_or_reinit_hook` in `object_detection_components/hooks.py`. For the denoising embedding, keep the trailing `padding_idx` row.
- Update the stale docstrings in that file and in `dfine_decoder.py` / `rtdetrv2_decoder.py`, which describe the truncate/pad behavior.
- Add `tests/_task_models/object_detection_components/test_hooks.py`. These hooks currently have no test coverage. Both decoders (`DFINETransformer`, `RTDETRTransformerv2`) register them, so cover matching, larger and smaller class counts.
- Note the behavior change in `CHANGELOG.md` under `### Changed`.

Happy to help anyone who picks this up.

**Comment by saitakarcesme** [comment]
I am working on a focused fix for the class-count mismatch behavior in the shared detection hooks, with CPU regression tests for both decoder families and unchanged matching-class behavior. This is a Codex-assisted contribution; I will include the validation results in the PR.

**Comment by papamamago** [comment]
## Fix: re-initialize classification heads when the class count differs

Thanks for the detailed report — the reproduction is exactly right. Confirmed on `5863b70`: with an 80-class COCO checkpoint loaded into a 3-class module, `dec_score_head.0.weight[0]` ends up byte-identical to the checkpoint's `person` row, and `denoising_class_embed` keeps the checkpoint's leading rows and padding row too.

### Root cause

Both mismatch branches of the object-detection hooks **reuse** checkpoint rows instead of re-initializing:

- `_reuse_or_reinit` (`hooks.py:168`): truncates when `num_classes_checkpoint > num_classes_module`, pads when smaller.
- `denoising_class_embed_reuse_or_reinit_hook` (`hooks.py:31`): same, plus it explicitly appends the *checkpoint's* `padding` row (`checkpoint_weight[-1:]`) in the truncate branch.

The det/seg hooks disagree: `eomt/hooks.py:85` already overwrites the checkpoint entry with `class_head_module.weight.detach().clone()`, and the `train_object_detection` docstring promises re-initialization. So only the object-detection path was silently inheriting the checkpoint's class ordering.

### Change

`src/lightly_train/_task_models/object_detection_components/hooks.py`

When the class counts differ, keep the module's own initialization instead of reusing checkpoint rows. When they match, behavior is unchanged (checkpoint reused as before).

```python
# _reuse_or_reinit
    num_classes_module = getattr(head_module, "out_features", None)
    if num_classes_module is None or num_classes_checkpoint == num_classes_module:
        return False

    # Keep the module initialization by overwriting the checkpoint weights with the
    # current parameter tensors.
    state_dict[weight_key] = head_module.weight.detach().clone()
    if checkpoint_bias is not None:
        state_dict[bias_key] = head_module.bias.detach().clone()
    return True
```

```python
# denoising_class_embed_reuse_or_reinit_hook
    # Keep the module initialization (including its trailing padding_idx row) by
    # overwriting the checkpoint weight with the current parameter tensor.
    state_dict[weight_key] = embed_module.weight.detach().clone()
```

The `denoising_class_embed` padding row is now taken from the module itself, so `padding_idx` is preserved as you requested.

Also:
- Updated the stale module/function docstrings that described the old truncate/pad behavior.
- Added `CHANGELOG.md` entry under `### Changed`.

The decoder headers in `dfine_decoder.py` / `rtdetrv2_decoder.py` only say the checkpoints are "adapted to a new `num_classes` at load time", which stays accurate, so I left them alone.

### Tests

Added `tests/_task_models/object_detection_components/test_hooks.py` (the hooks had no coverage). Which decoders register them — both `DFINETransformer` and `RTDETRTransformerv2` — is exercised through the same shared hook functions; the tests cover all three cases plus the no-op paths:

| Test | Case |
|---|---|
| `test_score_head_matching_num_classes_reuses_checkpoint` | equal counts → checkpoint reused |
| `test_score_head_fewer_classes_reinitializes_from_module` | ckpt 80 → module 3 → re-init |
| `test_score_head_more_classes_reinitializes_from_module` | ckpt 3 → module 10 → re-init |
| `test_score_head_single_head_module` | single (non-`ModuleList`) encoder head |
| `test_score_head_missing_keys_are_noop` | missing keys → untouched |
| `test_denoising_class_embed_matching_num_classes_reuses_checkpoint` | equal counts → reused |
| `test_denoising_class_embed_different_num_classes_reinitializes_with_padding` | diff counts → re-init, padding row preserved |
| `test_denoising_class_embed_missing_key_is_noop` | missing key → untouched |

**Result on the fixed tree:**

```
$ python -m pytest tests/_task_models/object_detection_components/test_hooks.py -q
........                                                                 [100%]
8 passed in 2.61s
```

**Counter-proof** — same tests against the unpatched source:

```
$ git stash   # revert src/, keep the new tests
$ python -m pytest tests/_task_models/object_detection_components/test_hooks.py -q
FAILED ...::test_score_head_fewer_classes_reinitializes_from_module
FAILED ...::test_score_head_more_classes_reinitializes_from_module
FAILED ...::test_score_head_single_head_module
FAILED ...::test_denoising_class_embed_different_num_classes_reinitializes_with_padding
4 failed, 4 passed in 2.37s
```

The four failures are exactly the re-initialization tests; the four that still pass are the matching-count and no-op cases, so the suite pins the bug rather than passing vacuously.

I also ran an independent standalone repro (`torch.equal(fresh, after)` → `True`, and `after[0]` no longer equals the COCO `person` row) and confirmed the module compiles (`py_compile`).

### Note on opening a PR

I can't open a PR for this repo right now — my account is currently blocked from forking (`You cannot fork this repository at this time`). The full patch is below and applies cleanly on `5863b70`. Happy to turn it into a PR as soon as that's resolved.

<details>
<summary>Full diff</summary>

```diff
diff --git a/CHANGELOG.md b/CHANGELOG.md
index xxx..yyy 100644
--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -36,6 +36,15 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
 
 ### Changed
 
+- Object detection fine-tuning on a checkpoint with a different number of classes now
+  re-initializes the classification head (both the score heads and the denoising class
+  embeddings) from the module's own initialization instead of truncating or padding the
+  checkpoint weights. Previously, fine-tuning a COCO checkpoint (80 classes) onto a
+  smaller class set silently started from the checkpoint rows at those indices — for
+  example the rows for `person`, `bicycle` and `car` when training on three classes — so
+  the initial bias depended on the checkpoint's class ordering. This now matches the
+  behavior of the EoMT segmentation heads and the documented
+  `train_object_detection` behavior.
 - Fine-tuning now requires `torchmetrics>=1.5` and fails with a clear error message if
   an older version is installed. This happens for example when SuperGradients is
   installed, as it requires `torchmetrics==0.8`. Classification, segmentation, and
diff --git a/src/lightly_train/_task_models/object_detection_components/hooks.py b/src/lightly_train/_task_models/object_detection_components/hooks.py
--- a/src/lightly_train/_task_models/object_detection_components/hooks.py
+++ b/src/lightly_train/_task_models/object_detection_components/hooks.py
@@ -13,9 +13,10 @@
 """Load-state-dict pre-hooks shared by RT-DETRv2 / D-FINE decoders.
 
 These hooks adapt a pretrained checkpoint to a module configured with a
-different number of classes, by truncating or zero-/re-initializing the
-classification-related weights so ``load_state_dict`` succeeds.
+different number of classes. The classification-related weights are either
+reused as-is (when the class counts match) or re-initialized from the module's
+own initialization (when they differ), so that ``load_state_dict`` succeeds
+without silently inheriting an arbitrary class ordering from the checkpoint.
 """
 
 from __future__ import annotations
@@ -39,10 +40,11 @@
     """Adjust denoising class embeddings when checkpoint has different number of classes.
 
-    If the checkpoint and module have different numbers of classes, this hook reuses
-    available weights and initializes missing ones from the module's initialization.
-    This allows loading checkpoints trained on different datasets.
+    If the checkpoint and module have different numbers of classes, this hook keeps the
+    module's own initialization instead of reusing the checkpoint's class rows. This
+    avoids silently inheriting an arbitrary class ordering from a checkpoint that was
+    trained on a different dataset. The trailing ``padding_idx`` row of the module is
+    always preserved.
 
     Args:
         module: The module being loaded.
@@ -64,36 +66,12 @@
     logger.info(
         f"Checkpoint has {num_classes_checkpoint - 1} classes, module expects "
-        f"{num_classes_module - 1} classes. Adjusting denoising class embeddings."
+        f"{num_classes_module - 1} classes. Reinitializing denoising class embeddings."
     )
 
-    device = embed_module.weight.device
-
-    # Last class is padding_idx
-    num_user_classes_checkpoint = num_classes_checkpoint - 1
-    num_user_classes_module = num_classes_module - 1
-
-    if num_classes_checkpoint > num_classes_module:
-        # Checkpoint has more classes: reuse checkpoint and discard excess
-        adjusted_weight = torch.cat(
-            [
-                checkpoint_weight[:num_user_classes_module].to(device),
-                checkpoint_weight[-1:].to(device),  # padding class
-            ],
-            dim=0,
-        )
-    else:
-        # Checkpoint has fewer classes: reuse checkpoint and initialize missing from
-        # module
-        adjusted_weight = torch.cat(
-            [
-                checkpoint_weight[:num_user_classes_checkpoint].to(device),
-                embed_module.weight[num_user_classes_checkpoint:].detach().clone(),  # type: ignore[index]
-            ],
-            dim=0,
-        )
-
-    state_dict[weight_key] = adjusted_weight
+    # Keep the module initialization (including its trailing padding_idx row) by
+    # overwriting the checkpoint weight with the current parameter tensor.
+    state_dict[weight_key] = embed_module.weight.detach().clone()
 
 
 def score_head_reuse_or_reinit_hook(
@@ -156,9 +134,10 @@
     """Adjust linear head weights/biases when checkpoint has different number of classes.
 
-    Enables loading checkpoints trained on different number of classes by either:
-    - Truncating weights if checkpoint has more classes (excess classes discarded)
-    - Padding weights if checkpoint has fewer classes (new classes initialized from module)
+    Enables loading checkpoints trained on a different number of classes by keeping the
+    module's own initialization whenever the class counts differ. Reusing (truncating or
+    padding) the checkpoint rows would silently bias the new classes towards whatever
+    classes happened to sit at those indices in the checkpoint.
 
     Args:
         head_module: The linear classification head module.
@@ -177,32 +156,9 @@
     if num_classes_module is None or num_classes_checkpoint == num_classes_module:
         return False
 
-    device = head_module.weight.device
-
-    if num_classes_checkpoint > num_classes_module:
-        # Checkpoint has more classes: truncate to module's expected size.
-        adjusted_weights = checkpoint_weight[:num_classes_module, :].to(device)
-        if checkpoint_bias is not None:
-            adjusted_biases = checkpoint_bias[:num_classes_module].to(device)
-    else:
-        # Checkpoint has fewer classes: pad with module's initialized weights.
-        adjusted_weights = torch.cat(
-            [
-                checkpoint_weight.to(device),
-                head_module.weight[num_classes_checkpoint:].detach().clone(),  # type: ignore[index]
-            ],
-            dim=0,
-        )
-        if checkpoint_bias is not None:
-            adjusted_biases = torch.cat(
-                [
-                    checkpoint_bias.to(device),
-                    head_module.bias[num_classes_checkpoint:].detach().clone(),  # type: ignore[index]
-                ],
-                dim=0,
-            )
-
-    state_dict[weight_key] = adjusted_weights
+    # Keep the module initialization by overwriting the checkpoint weights with the
+    # current parameter tensors.
+    state_dict[weight_key] = head_module.weight.detach().clone()
     if checkpoint_bias is not None:
-        state_dict[bias_key] = adjusted_biases
+        state_dict[bias_key] = head_module.bias.detach().clone()
     return True
```

</details>

If you'd prefer the heads to raise a `DenckringError`-style explicit error instead of silently re-initializing, say the word and I'll switch it — but re-init matches the EoMT hooks and the documented behavior, so that's what I went with.
